### PR TITLE
 Fixed PHP 5.3 compatibility

### PR DESCRIPTION
**Pull request by enumag**
@dg Pick whatever commit(s) you want, or just close this PR if you don't want any of it. I can squash it if you wish.

### CODE REVIEW

File: src/Reflection/Helpers.php [hunk 1]
```diff
@@ -21,14 +21,17 @@ class Helpers
 	 * @return \ReflectionClass
 	 * @internal
 	 */
-	public static function getDeclaringClass(\ReflectionProperty $prop)
+	public static function getDeclaringClass(\ReflectionProperty $property)
 	{
-		foreach ($prop->getDeclaringClass()->getTraits() as $trait) {
-			if ($trait->hasProperty($prop->getName())) {
-				return self::getDeclaringClass($trait->getProperty($prop->getName()));
+		if (method_exists('ReflectionClass', 'getTraits')) {
```
Comment (Majkl578) [inline src/Reflection/Helpers.php]: Why not `PHP_VERSION_ID >= 50400`?

Comment (enumag) [inline src/Reflection/Helpers.php]: Because I could't find such check anywhere in Nette.

Comment (enumag) [inline src/Reflection/Helpers.php]: Yeah that's better, thanks.

**Comment by fprochazka** [comment]
This is a nice refactoring. :+1:

**Comment by JanTvrdik** [comment]
Only it should have been a single commit with a meaningful commit message. But maybe @dg gonna force push master, so it does not matter :smile:

**Comment by fprochazka** [comment]
@JanTvrdik

**Comment by enumag** [comment]
@JanTvrdik I split it intentionally because I don't know which changes will @dg want to merge. When he will tell me, I will squash it. Or he can do it himself. :-)

**Comment by dg** [comment]
First one is good enough, isnt it?

**Comment by enumag** [comment]
Yes, the other commits are just refactoring suggested by @JanTvrdik (and I agree with him).

**Comment by dg** [comment]
Removing duplicate calling inside foreach is nonsense. But do not fix it, this is unnecessary optimization, I'll take first commit.

**Comment by JanTvrdik** [comment]
@dg Why nonsense? Method call is more expensive than variable fetch. But this is not about performance, this is about readability and being DRY. Also I don't like (visually) `()))`. Obviously these are not strong arguments but it is not a nonsense and it is worth being considered.

**Comment by dg** [comment]
Storing member to local variable is not about DRY, accessing the same member is not repeating. Readability is matter of opinion, for me is getName() fine. It is nonsense from performance view, because variable is created inside foreach.

**Comment by JanTvrdik** [comment]
> It is nonsense from performance view, because variable is created inside foreach.

Variable is always created. Either named (proposed solution) or temporary (current solution). Also in real-world use cases it would be likely very common that the `foreach` cycle would have zero or one iteration therefore the created variable would not be thrown away. However as state the proposed changed should not be considered from performance point of view (as the difference is negligible) but as a readability improvement. That is (of course) matter of personal taste.

**Comment by enumag** [comment]
thx @dg :-)

**Comment by dg** [comment]
God, again we are dealing with %&$ :-)

On my computer, var1() is a little bit faster than var2(), believe me. 

```
function var1()
{
	foreach (range(1,1000) as $i) {
		if ($i === x()) {
			return x();
		}
	}
}

function var2()
{
	foreach (range(1,1000) as $i) {
		$var = x();
		if ($i === $var) {
			return $var;
		}
	}
}
```

**Comment by JanTvrdik** [comment]
@dg Not if `x()` returns low value. That `var2` becomes faster.

**Comment by fprochazka** [comment]
`<offTopic class="well-intentioned-advice">`

If you write things like "nonsense" people get defensive (somebody even goes to ofensive) and try to defend themselfs - nobody want's to be viewed as "that guy who wrote that nonsense pullrequest". That's exactly the attitude that will scare away contributors.

I don't think this discussion would have ever happened if you were to choose your words more wisely. I'm not saying you **have to** be politicaly correct, but you can't be wrong by beeing more polite. @dg you've always been the voice of reason, please be a little bit more patient and less agresive. It will make everyone happy. 

`</offTopic>`

**Comment by dg** [comment]
Aggressive? Oh :-(

**Comment by JanTvrdik** [comment]
Looking at the code once again… Why noone proposed to the `getName()` called out of the `foreach`? :smile: 

~~~php
$name = $prop->getName();
foreach ($prop->getDeclaringClass()->getTraits() as $trait) {
    if ($trait->hasProperty($name)) {
        return self::getDeclaringClass($trait->getProperty($name));
    }
}
~~~

**Comment by fprochazka** [comment]
`<ot>`
@dg Written text doesn't have voice intonation, so it's really easy to write something well intentioned, that sounds agresive or arogant. [I had my share of those](https://filip-prochazka.com/blog/uz-nebudu-radit-zacatecnikum-predavam-stafetu) so I try to be careful about what I write.
`</ot>`

**Comment by dg** [comment]
@JanTvrdik eh? „Removing duplicate calling **inside foreach** is (aggressive word).“

**Comment by enumag** [comment]
@JanTvrdik Ups, my mistake. :-/

**Comment by dg** [comment]
@fprochazka česky řeknu: „hele vytvářet proměnnou uvnitř cyklu je nesmysl,“ cítíš se tím uražen? Já bych to tak necítil. A fakt se snažím být slušný (a že se často musím opravdu snažit). Ale tohle mi připadá jako přecitlivělost. Hracha na vás ;-)

**Comment by fprochazka** [comment]
@dg necítím :) česky mi to zní úplně jinak :)

**Comment by dg** [comment]
@enumag But please do not fix it, this is really unnecessary optimization. Use-case is cached.

**Comment by JanTvrdik** [comment]
@dg Oh, now I finally understand what "But do not fix it" meant in https://github.com/nette/reflection/pull/6#issuecomment-68866709 :smile: You have known that it is a nonsense the whole time.

**Pull request closed without merging**